Add blog post about collecting diagnostic data (#476)
strimzi/strimzi.github.io: _posts/2025-03-17-collecting-diagnostic-data.md

---
layout: post
title: "Collecting diagnostic data"
date: 2025-03-17
author: federico_valeri
---

Debugging distributed systems like Apache Kafka can be challenging.
Without access to the right diagnostic data, identifying the root cause of an issue can become a time-consuming process.

> Strimzi, the Kubernetes-native Kafka operator, provides a [bash script](https://github.com/strimzi/strimzi-kafka-operator/blob/0.45.0/tools/report.sh) to collect cluster artifacts, including logs and configurations.

Reviewer: As it is a not-released artifact, I wonder if linking to the main branch would be better. But it might be also worse if we one day decide to remove or move it. So up to you.

Author: I'm using the tag here exactly to avoid the risk of code changes.

> Sometimes, this is not enough and we need long-term storage to collect custom and advanced diagnostic data such as heap dumps and flame graphs.

Reviewer: Do you really need long term storage? Or do you just need high-capacity storage where you can store the data and later copy them out for analysis?

Author: Probably "persistent" is a better term for what I wanted to say, but I will also add high-capacity.

---
layout: post
title: "Collecting diagnostic data"
date: 2025-03-17
author: federico_valeri
---

> Debugging distributed systems like Apache Kafka can be challenging.

Reviewer 2: ```suggestion
Debugging distributed systems, such as Apache Kafka, can be challenging.
```

Without access to the right diagnostic data, identifying the root cause of an issue can become a time-consuming process.
Strimzi, the Kubernetes-native Kafka operator, provides a [bash script](https://github.com/strimzi/strimzi-kafka-operator/blob/0.45.0/tools/report.sh) to collect cluster artifacts, including logs and configurations.

> Sometimes, this is not enough and we need persistent high-capacity storage to collect custom and advanced diagnostic data.

Reviewer 2: shall we explain more about when the default collection is insufficient? When we need more details or longer term retention?

Fortunately, Strimzi recently introduced the **additional volumes** feature that makes it easier to capture these artifacts by leveraging Kubernetes-native mechanisms such as persistent volumes.

<!--more-->

### The need for diagnostic data in Kafka troubleshooting

Kafka clusters can encounter a range of issues, from resource exhaustion to unexpected application crashes.
When these issues arise, collecting relevant diagnostic data is essential for effective debugging.
Without this information, developers are left guessing, often leading to unnecessary downtime and prolonged resolutions.

> Common diagnostic artifacts needed for debugging Kafka and Java application in general include:

Reviewer 2: ```suggestion
Common diagnostic artifacts needed for debugging Kafka and Java applications in general include:
```

- **Thread dumps**: To inspect thread states and potential deadlocks.
- **Heap dumps**: To analyze memory usage and detect memory leaks.
- **JVM metrics**: To monitor CPU, memory, and GC activity.
- **Log files**: To track application errors and warnings.
- **Flame graphs**: To visually identify performance bottlenecks.

### Introducing Strimzi additional volumes

The additional volumes feature, introduced via [proposal 75](https://github.com/strimzi/proposals/blob/main/075-additional-volumes-support.md), allows users to define extra storage volumes in all operands.
This enhancement supports different types of volumes, but here we are interested in `PersistentVolumeClaims` to durably store and retrieve any kind of diagnostic data under the `/mnt` mount point.

> > Adding a custom volume will roll the pods, which can make it difficult to capture an issue that has already occurred.

Reviewer 2: ```suggestion
> Adding a custom volume triggers pod restarts, which can make it difficult to capture an issue that has already occurred.
```

> If the issue cannot be easily reproduced in a test environment, configuring the volume in advance could help avoid the pod restarts when you need them most.

With additional volumes, users can:

- Mount persistent or ephemeral storage for logs, dumps, or other debugging artifacts.
- Capture JVM dumps without modifying the default storage configuration.
- Enable better debugging workflows by keeping artifacts accessible even after a pod restart.

> Debugging locally can often be easier and faster.
> However, there are several reasons why you might still want to do this kind of debugging on Kubernetes:

Reviewer 2: ```suggestion
Although debugging locally is often easier and faster, certain issues are best diagnosed within a Kubernetes environment due to the following factors:
```

- **Environment parity**: Some issues only manifest in Kubernetes due to factors like networking, resource limits, or interactions with other components.
  Debugging in the actual deployment environment can help reproduce and diagnose these problems.
- **Configuration differences**: Even if you try to match your local setup to the Kubernetes configuration, subtle differences (e.g. service discovery, security settings, or operator-managed logic) might lead to different behavior.

### Example: capturing heap dumps with additional volumes

A practical example of how additional volumes simplify debugging is demonstrated in the following procedure.
The scenario involves collecting heap dumps from a Kafka broker when observing excessive memory consumption.

> Memory leaks can sometimes lead to out of memory exceptions (OOME) and service disruption.

Reviewer 2: ```suggestion
Memory leaks can sometimes lead to an `OutOfMemoryError` (OOME) and service disruption.
```

> > Taking a heap dump is a heavy operation that makes the Java application hangs.

Reviewer 2: ```suggestion
> Taking a heap dump is a heavy operation that can cause the Java application to hang.
```
